pkg/kubelet/kubelet.go

```diff
@@ -283,8 +283,19 @@ func makePodSourceConfig(kubeCfg *kubeletconfiginternal.KubeletConfiguration, ku
 	}
 
 	if kubeDeps.KubeClient != nil {
-		klog.InfoS("Watching apiserver")
-		config.NewSourceApiserver(kubeDeps.KubeClient, nodeName, cfg.Channel(kubetypes.ApiserverSource))
+		klog.InfoS("Waiting for node sync before watching apiserver pods")
+		go func() {
+			for {
+				if nodeHasSynced() {
+					klog.V(4).InfoS("node sync completed")
+					break
+				}
+				time.Sleep(waitForAPIServerSyncPeriod)
+				klog.V(4).InfoS("node sync has not completed yet")
+			}
+			klog.InfoS("Watching apiserver")
+			config.NewSourceApiserver(kubeDeps.KubeClient, nodeName, cfg.Channel(kubetypes.ApiserverSource))
```

Comment:
I think that inserting this source asynchronously (potentially seconds or minutes after the kubelet started) means that `config.SeenAllSources()` can return true too early, not understanding that an API source will be inserted later.

That can cause extremely disruptive behavior on kubelet restart when there are existing running containers, since many cleanup behaviors gate on `sourcesReady.AllReady()` to know if they should start cleaning up existing containers/volumes/etc.

Reply:
makePodSourceConfig() has to return a source config that knows it has an API config source, but which will not propagate pods read from that source until the node sync has completed

Reply:
we can call NewSourceApiserver as part of the main thread, but async / delay it's Run on the reflector:
https://github.com/kubernetes/kubernetes/blob/a42ff247afeb2f2d7aaf1d28a7cb171d722fb527/pkg/kubelet/config/apiserver.go#L46
until the node cache has synced.

[1] i assume not running the reflector would delay the SET message [2] which would mean config.SeenAllSources would return false. 

[1]
https://github.com/kubernetes/kubernetes/blob/58814dbe3fc13602fb8a9ac67f9e046c9eb09326/pkg/kubelet/config/config.go#L91-L92

[2]
https://github.com/kubernetes/kubernetes/blob/a42ff247afeb2f2d7aaf1d28a7cb171d722fb527/pkg/kubelet/config/apiserver.go#L43

and that would solve the sourcesReady.AllReady() problem too?

Reply:
that sounds right to me